### PR TITLE
LPS-137762 Use ballooneditor in blogs coverImageEditor

### PR DESCRIPTION
**Pull request by diegonvs**
# Current status

https://user-images.githubusercontent.com/7663513/130699748-dc4b1d03-33dc-4c12-9909-8242f570d7de.mov

_Watch while listening to https://www.youtube.com/watch?v=yg8116aeD7E_

# Things to do:

- [x]  Controlling the editor

~~Blogs requires controlling the editor. It uses a primitive way that is querying for the element in the DOM and then perfoming actions. These actions can be cleaning the content inside the editor and getting the content inside it.~~

~~It can be accomplished because the editor apparently is wrapped by an [AUI plugin](https://github.com/liferay/liferay-portal/blob/de6379116225b1d2c9aee14d0a37d20c8660b0cc/modules/apps/frontend-editor/frontend-editor-alloyeditor-web/src/main/resources/META-INF/resources/js/alloyeditor.js#L27).~~

See:

Screen Shot 2021-08-24 at 17 41 48

Solved using `CKEditor.instances` global and setData and getData

- [x] Render just one link button in textToolbar

~~When showing toolbar buttons, some of them are disabled. It's correct because We are just allowing link button but in the legacy editor We are not showing those buttons.~~

Screen Shot 2021-08-24 at 17 52 51

Solved passing empty string as values for toolbar(video|image...), defining toolbarlink and toolbartext with just TextLink button.

## Usage design question:

I can still working with *Contributor things but it's very difficult to handle things there. Can I work with instead of using `liferay-editor:editor`, use a `react:component` just importing the BalloonToolbar component and extending the defaults there using a JSON?

I can pass empty strings/arrays to properties in *Contributor.java files but it can be weird for people who never touched editor things before.

- [x] Wrong styles

~~We're loading moono-lisa skin instead of moono-lexicon.~~

This is not completely true, because We're loading `moono-lisa` and it's correct to make AlloyEditor in blogs body editor work but Weren't loading `balloontoolbar.css` from `moono-lexicon` skin. I tried to add a new entry in `contentsCss` property, resolving a path to loading this specific file and it don't worked. Also, I saw an usage of AlloyEditor in PageEditor and I saw a property called `skin` and it's being setted as `skin: "moono-lisa"`. I decided to pass `skin: "moono-lexicon"` but it don't worked 😢 

The example from ckeditor-sample it's working fine and it's loading this style but it's not the same for this blogs thing, even using BaseCKEditorConfig(yes, I know in this PR We're using a base config for a generic editor instead of a CKEditor contributor Preset) as a preset in Contributor part

I saw `2021-08-25 20:16:00.903 WARN  [http-nio-8080-exec-9][code_jsp:161] {code="404", msg="Not Found", uri=/o/frontend-editor-ckeditor-web/ckeditor/skins/moono-lexicon/balloontoolbar.css}` when trying to require the style and I forced pointed to the correct file and it still not working 

Opened: https://issues.liferay.com/browse/LPS-138200

- [x] SPA failing

When saving and then getting back to the `coverImageCaptionEditor` styles are missing.

Note: When refreshing the page, everything works fine.

My guess: I think We could use senna-data-track when appending CKE styles. It needs to be investigated

Screen Shot 2021-08-24 at 19 29 48

Opened: https://issues.liferay.com/browse/LPS-138203

### CODE REVIEW

**Comment by diegonvs** [comment]
@markocikos, all green. Can I fwd?

File: modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java
Comment (markocikos) [inline modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java:35]: If we are setting editor name here, we don't have to set it in JSP. It's probably better to leave the update in JSP, because it is consistent with setting in JSP when editor is not hardcoded. Normally it's not hardcoded, but set in `portal-ext.properties`.

File: modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java
Comment (markocikos) [inline modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java:59]: `else` block is not covered.

File: modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java
Comment (markocikos) [inline modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java:72]: Doesn't this override text toolbar with just one button? That can't be right.

File: modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/edit_entry.jsp
Comment (markocikos) [inline modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/edit_entry.jsp:145]: We cannot just temporarily remove a feature from product team component. We can, if there is a really good reason for it, but there is none in this case. We need to implement the [source mode support](https://issues.liferay.com/browse/LPS-137763) first.

File: modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js
Comment (markocikos) [inline modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js:320]: I wonder if this is safe to do. The original setHTML seems to have [handled some edge cases](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/ckeditor.jsp#L302-L325) , beyond happy days setting of data.

**Review by markocikos — CHANGES_REQUESTED** [review]
@diegonvs See comments inline. The biggest blocker for this PR is removal of the source feature. This is a migration of an actual live product team component. If a PR makes the actual switch, we don't have the luxury of forwarding a half-finished solution. We can only do that while the only usage is on a sample module.

**Comment by markocikos** [comment]
@diegonvs Have in mind, this should be forwarded to the product team first. Make sure that we have complete feature parity with AE, be prepared for discussion about styles. The stability period for 7.4 is about to start, but this probably a good thing for the epic. We can have a longer discussion with the product team and polish.

File: modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java
Comment (diegonvs) [reply modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java:72]: It was necessary to achieve the "only-button" toolbar

See the current master:

Screen Shot 2021-08-30 at 12 35 43

File: modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js
Comment (diegonvs) [reply modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js:320]: It looks interesting. Can I create another function that handles the original behavior of setHTML from AUI's component?

File: modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java
Comment (diegonvs) [reply modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java:35]: Nice! I removed `"editor.name=ballooneditor",` from OSGi properties and It works fine.

File: modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java
Comment (diegonvs) [reply modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java:59]: Yes. I added `,stylescombo,videoembed` as plugins to be removed there too :)

**Comment by diegonvs** [comment]
> Have in mind, this should be forwarded to the product team first. Make sure that we have complete feature parity with AE, be prepared for discussion about styles.

Ok, thanks for reminding :)

I applied/commented the suggestions and I left a question here: https://github.com/liferay-frontend/liferay-portal/pull/1397#discussion_r698595278.

File: modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js
Comment (diegonvs) [reply modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js:320]: I'm waiting a definition for it before opening a new PR in blogs team(maybe t-lima?) repo.

Comment (markocikos) [reply modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js:320]: > Can I create another function that handles the original behavior of setHTML from AUI's component?

@diegonvs Sorry for the late reply!

What AUI component do you mean? If we are adding handling of some general edge cases of updating date, this should be done internally to Balloon Editor component. Now, `setHTML` is a very ugly API. I think what you are doing in this PR is much better, using vanilla CKEditor API. Imo we should:
1. Think if there is a better way to reference BE instance, to update its data. Probably not, but it's something to have in mind.
1. Analyse if there is anything in `setHTML` that we actually need. Are those edge cases still valid? It maybe be worth just letting PR as it is, and fix edge cases if any later.
1. If there is something we need to change, maybe there is a way to implement this with standard API in BE. I'm thinking exposed lifecycle moments  like `onChange`, `onSetData`, `onBeforeLoad`, `onInstanceReady` etc.

File: modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java
Comment (markocikos) [reply modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java:72]: That looks so weird :smile: . But, if their product folk like it, :man_shrugging:

File: modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js
Comment (markocikos) [reply modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js:320]: >maybe t-lima?

Yep, Lima is responsible for Blogs. I'm using [this](https://docs.google.com/spreadsheets/d/1egrstgNEMPf3KQT_KNpaemQ8vFMG5fEHGAmUOhP_UqU/edit#gid=0) to check. There is probably an updated version somewhere in Confluence :sweat_smile:

Comment (diegonvs) [reply modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js:320]: > What AUI component do you mean?

https://github.com/liferay/liferay-portal/blob/3bc9e133e07ebff6c1889ca834c49419570894f3/modules/apps/frontend-editor/frontend-editor-alloyeditor-web/src/main/resources/META-INF/resources/js/alloyeditor.js this component but I understood wrongly, it is defined in the JSP as you mentioned.

**Answering**

1. Think if there is a better way to reference BE instance, to update its data. If it's invoked with tag, probably not. I'm thinking something like propsTransformer, but that may be an overkill. It's something to have in mind.

> If it's invoked with tag

Yes, I'm not creative for it but We can improve by creating a class that receives the namespace and the editorName for handling those things. However, it can be take a long time to implement and test to see side effects. 

> Side note: I never took a look deeply in those things that `ckeditor.jsp` contains and I'm horrified (as We have in brazilian portuguese: kkkkkkkkcrying)

Comment (diegonvs) [reply modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js:320]: 2. Analyse if there is anything in setHTML that we actually need. Are those edge cases still valid? It maybe be worth just letting PR as it is, and fix edge cases if any later.

This piece of code:

```js
if (instanceDataReady) {
    ckEditorInstance.setData(data);
}
else {
    instancePendingData = data;
}
```

make me worry about it. I prefer do not touch

Comment (diegonvs) [reply modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js:320]: 3. If there is something we need to change, maybe there is a way to implement this with standard API in BE. I'm thinking exposed lifecycle moments like onChange, onSetData, onBeforeLoad, onInstanceReady etc.

I thing those callbacks can be easily replaced with propsTransformer API but it will take more and more time to made these changes. Does make sense adding another story/epic for it?

Comment (diegonvs) [reply modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js:320]: Summarizing changes:

1. Create a class to interact with CKEditor internals and avoid defining things in JSP side
2. Migrate current editor's backend lifecycle to propsTransformer pattern

Does make sense implementing it for this migration? I can spend a time to see if the current way it's worth just to cover this task

Comment (diegonvs) [reply modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js:320]: I can't create just for setHTML and getHTML because it depends in external things defined in JSP context.

For the ones who don't understood the problem yet and for me to make a question to myself:

We have a control for editors in the window:

Screen Shot 2021-09-06 at 19 08 22
<legend>Where each editor instance is represented in a window property.

and when trying to bring it to our context in a "boy-scouting" way, just creating the necessary functions it's doesn't working 😢  

I spent a time today trying to make things work as I already presented the style but I couldn't yet. Am I in a wrong direction? I need some guidance here /cc @nhpatt @markocikos

~~I was thinking about `extends EventEmitter` as we were doing in the past to have other methods like `dispose`~~

Comment (diegonvs) [reply modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js:320]: > I spent a time today trying to make things work as I already presented the style but I couldn't yet. Am I in a wrong direction?

Which can be translated in: I can be thinking in an "over-engineered" way

Comment (markocikos) [reply modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js:320]: It is probably best that we have the simplest solution for now, covering edge cases if and when they happen. That would be the solution in this PR, `CKEDITOR.instances[name].setData()`. Some of these cases may no longer be valid, like having to [handle setting data before instance is ready](https://github.com/liferay-frontend/liferay-portal/pull/1397#discussion_r702132029). If we reach a case where this is not enough, we can discuss what option for API layer on top makes the most sense.

**Comment by markocikos** [comment]
@diegonvs please force-push the latest changes

File: modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js
Comment (diegonvs) [reply modules/apps/blogs/blogs-web/src/main/resources/META-INF/resources/blogs/js/blogs.js:320]: > Some of these cases may no longer be valid, like having to handle setting data before instance is ready.

Yes, It makes sense. 

> If we reach a case where this is not enough, we can discuss what option for API layer on top makes the most sense.

omg

**Comment by diegonvs** [comment]
> @diegonvs please force-push the latest changes

Doing it.

**Comment by diegonvs** [comment]
ci:test:sf

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/css/main.scss
Comment (markocikos) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/css/main.scss:1]: Do we need a separate file for `balloontoolbar` stlyes?

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/css/_balloontoolbar.scss
Comment (markocikos) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/css/_balloontoolbar.scss:6]: Minor SF: there should be a blank row between sibling classes and between styles and first class.

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/css/_balloontoolbar.scss
Comment (markocikos) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/css/_balloontoolbar.scss:19]: Minor SF: if there is only one child, you can collapse. In this case, to `span:last-child`

File: modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java
Comment (markocikos) [inline modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java:62]: `magicline,stylescombo,videoembed` should probably be in a variable, so we don't make the same mistake because of duplication.

**Review by markocikos — CHANGES_REQUESTED** [review]
@diegonvs  Few comments inline, minor SF stuff.

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/css/main.scss
Comment (diegonvs) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/css/main.scss:1]: I think not. Let's merge it :p

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/css/_balloontoolbar.scss
Comment (diegonvs) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/css/_balloontoolbar.scss:19]: Makes sense! Changing!

File: modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java
Comment (diegonvs) [reply modules/apps/blogs/blogs-editor-configuration/src/main/java/com/liferay/blogs/editor/configuration/internal/BlogsCoverImageCaptionEditorConfigContributor.java]: done!

**Comment by diegonvs** [comment]
ci:test:sf

**Comment by diegonvs** [comment]
ci:test:relevant

**Review by markocikos — APPROVED** [review]
LGTM! If you like, feel free to forward to the product team, or ping them to review it here. They can review it, but this should definitely be on hold until 7.4 release.

**Comment by diegonvs** [comment]
Sent https://github.com/liferay-frontend/liferay-portal/pull/1397

**Pull request closed without merging**
